### ISSUE TITLE
objects inside abstract objects with the same name can't be build

### ISSUE THREAD

**Issue by maxonfjvipon** [objectionary/eo#1807]
I have 2 almost identical unit tests:
```
+alias org.eolang.collections.range
+alias org.eolang.collections.list
+alias org.eolang.hamcrest.assert-that
+home https://github.com/objectionary/eo-collections
+junit
+package org.eolang.collections
+version 0.0.0

[] > simple-range
  range > rng!
    []
      [i] > build
        i > @!
        build (@.plus 1) > next
      build 1 > @
    10
  assert-that > @
    list rng
    $.equal-to
      list
        * 1 2 3 4 5 6 7 8 9

[] > range-with-floats
  range > rng!
    []
      [i] > build
        i > @!
        build (@.plus 0.5) > next
      build 1.0 > @
    5.0
  assert-that > @
    list rng
    $.equal-to
      list
        * 1.0 1.5 2.0 2.5 3.0 3.5 4.0 4.5
```
In both of them I have an object `build` inside abstract objects
When I do `mvn clean test` I expect everything works fine but I get: 

_java.io.IOException: Failed while trying to save to /Users/maxonfjvipon/code/java/eo-collections/target/eo-test/04-pull/simple-range$t0$a0$build range-with-floats$t0$a0$build.eo_

_Failed to execute goal org.eolang:eo-maven-plugin:0.29.0:assemble (test-compile) on project eo-collections: 'org.eolang.maven.AssembleMojo@4a1fb55b' execution failed: java.lang.IllegalStateException: org.apache.maven.plugin.MojoFailureException: 'org.eolang.maven.PullMojo@21227342' execution failed: java.io.IOException: Failed while trying to save to /Users/maxonfjvipon/code/java/eo-collections/target/eo-test/04-pull/simple-range$t0$a0$build range-with-floats$t0$a0$build.eo: Server returned HTTP response code: 400 for URL: https://raw.githubusercontent.com/objectionary/home/0cbb782f2fb06ee966fdf73dd289391d231bc59f/objects/simple-range$t0$a0$build range-with-floats$t0$a0$build.eo_

If I rename object `build` in the second test - everything works fine

**Comment by maxonfjvipon** [comment]
@mximp here it goes: [stacktrace.txt](https://github.com/objectionary/eo/files/11000630/stacktrace.txt)

**Comment by mximp** [comment]
@maxonfjvipon are you building this within `eo-collections` module? Does the module builds successfully without those tests?

**Comment by maxonfjvipon** [comment]
@mximp yes, within `eo-collections`. Note that the `range` object has already been merged, so it worked WITH and WITHOUT those tests (it does not build on my machine right now for some reason neither WITH nor WITHOUT those tests because of broken pipeline). I just gave different names to `build` object in different tests (see [this PR](https://github.com/objectionary/eo-collections/pull/141/files))

I believe it's not about these specific tests, it's about some problem with giving the same names to the inner objects which are placed in different outer objects (like in example above)

**Comment by mximp** [comment]
@maxonfjvipon do you have representative example of the problem outside `eo-collections`?

**Comment by mximp** [comment]
The issue happens due to combination of two steps in EO optimization logic: `abstracts-float-up` and `fix-missed-names`. The fist one makes objects float up to first level of hierarchy setting original name. The second uses original name to find relevant bases. If two abstracts have the same name it will make the second optimization to identify two related objects and use them for forming `base` attribute. This leads to names like this:
```
<o base="bad-names$t1$t0$t1$t2$a0$build1$build bad-names$t1$t0$t1$t2$a1$build"
             line="22"
             loc="Φ.bad-names.φ.α0.α0.α1.α1.build.next"
             name="next"
             pos="14">
```

**Comment by mximp** [comment]
@yegor256 This affect quite fundamental logic: the semantic of identifying object by names in EO programs. Currently you can define an object wherever in the program and it can be referenced from any location. But seems this forces the programmer to name objects uniquely.  
Do you think we need to change the behaviour?

**Comment by yegor256** [comment]
@mximp I believe, `abstracts-float-up` must give proper _unique_ names to objects, when they float up. If this doesn't happen, it's a bug in this XSL

**Comment by mximp** [comment]
@yegor256 the XSL gives proper unique names to the objects. The problem is that it doesn't update all the references and only keep old name within `original-name`. References are updated asynchronously in another XSL based on `original-name`, when the information about nesting is lost already.
To avoid this `abstracts-float-up` must update all the references correctly once it give new unique name to an object.
But again this will change the semantics of the language: scope of object visibility will be narrowed.

**Comment by mximp** [comment]
@maxonfjvipon the issue has been resolved. Please check and close it.

**Comment by maxonfjvipon** [comment]
@mximp Seems to work! Thanks!
